File: Vision.m
function indices = Vision(que,egocar,k)
%VISIONTEST Summary of this function goes here
%   Detailed explanation goes here
len = length(que);
R = 50;
indices = [];
EgoHeading = atan2(egocar.realpose(2)-egocar.prerealpose(2),egocar.realpose(1)-egocar.prerealpose(1));
for i=1:1: len
    if i == k
        continue
    else
        RelativeHeading = EgoHeading - atan2(-egocar.prerealpose(2)-0.01+que{i}.realpose(2),-egocar.prerealpose(1)-0.01+que{i}.realpose(1));
        if (egocar.realpose(1)-que{i}.realpose(1))^2+(egocar.realpose(2)-que{i}.realpose(2))^2 <= R^2
            if -pi/8 <= RelativeHeading &&   RelativeHeading <= pi/8
                indices = [indices;i];
            end
        end        
    end

end

end

File: Carla_Sumo_Code/vision.py
import numpy as np

def vision(que, egocar, k):
    # Constants
    precision_rate = 0.1  # Set your value for 'precision_rate' here
    R = 50

    indices = []
    ego_heading = np.arctan2(egocar['realpose'][1] - egocar['prerealpose'][1], egocar['realpose'][0] - egocar['prerealpose'][0])

    for i, other_car in enumerate(que, start=1):
        if i == k:
            continue
        else:
            relative_heading = ego_heading - np.arctan2(-egocar['prerealpose'][1] - 0.01 + other_car['realpose'][1],
                                                         -egocar['prerealpose'][0] - 0.01 + other_car['realpose'][0])
            distance_squared = (egocar['realpose'][0] - other_car['realpose'][0]) ** 2 + \
                               (egocar['realpose'][1] - other_car['realpose'][1]) ** 2

            if distance_squared <= R ** 2 and -np.pi / 8 <= relative_heading <= np.pi / 8:
                #if perception_precision >= 1 - precision_rate:
                indices.append(i)

    return indices


# Example usage
que_data = [
    {'realpose': [1.0, 2.0]},  # Replace with your actual data
    {'realpose': [3.0, 4.0]},  # Replace with your actual data
    # ... add more dictionaries as needed
]

egocar_data = {'realpose': [5.0, 6.0], 'prerealpose': [3.0, 4.0]}  # Replace with your actual data
k_data = 2  # Replace with your actual data

indices_result = vision(que_data, egocar_data, k_data)
print("Indices:", indices_result)
